From 9e965c23b1b8adfcf38059c9f32319a2dbd85f15 Mon Sep 17 00:00:00 2001
From: Alejandro del Castillo <alejandro.delcastillo@ni.com>
Date: Wed, 22 Feb 2017 17:32:10 -0600
Subject: [PATCH 3/4] libopkg: add option verbose_status_file

Enabling this option makes opkg store all the package information in
the status file, instead of just a subset. This is desirable to make
such fields available offline, when there is no connection to the repo.
Disabled by default.

Signed-off-by: Alejandro del Castillo <alejandro.delcastillo@ni.com>
---
 libopkg/opkg_conf.c |  1 +
 libopkg/opkg_conf.h |  1 +
 libopkg/pkg.c       | 13 +++++++++++++
 libopkg/pkg_parse.c |  8 ++++++--
 4 files changed, 21 insertions(+), 2 deletions(-)

diff --git a/libopkg/opkg_conf.c b/libopkg/opkg_conf.c
index 7b63ad2..b9cc303 100644
--- a/libopkg/opkg_conf.c
+++ b/libopkg/opkg_conf.c
@@ -89,6 +89,7 @@ static opkg_option_t options[] = {
     {"overwrite_no_owner", OPKG_OPT_TYPE_BOOL, &_conf.overwrite_no_owner},
     {"combine", OPKG_OPT_TYPE_BOOL, &_conf.combine},
     {"cache_local_files", OPKG_OPT_TYPE_BOOL, &_conf.cache_local_files},
+    {"verbose_status_file", OPKG_OPT_TYPE_BOOL, &_conf.verbose_status_file},
 #if defined(HAVE_GPGME)
     {"gpg_dir", OPKG_OPT_TYPE_STRING, &_conf.gpg_dir},
     {"gpg_trust_level", OPKG_OPT_TYPE_STRING, &_conf.gpg_trust_level},
diff --git a/libopkg/opkg_conf.h b/libopkg/opkg_conf.h
index dfa2c57..fcac8dd 100644
--- a/libopkg/opkg_conf.h
+++ b/libopkg/opkg_conf.h
@@ -118,6 +118,7 @@ typedef struct opkg_conf {
     int combine;
     int cache_local_files;
     int host_cache_dir;
+    int verbose_status_file;
 
     /* ssl options: used only when opkg is configured with '--enable-curl',
      * otherwise always NULL or 0.
diff --git a/libopkg/pkg.c b/libopkg/pkg.c
index 52194d1..7f5d0bb 100644
--- a/libopkg/pkg.c
+++ b/libopkg/pkg.c
@@ -850,9 +850,22 @@ void pkg_print_status(pkg_t * pkg, FILE * file)
     pkg_formatted_field(file, pkg, "Replaces");
     pkg_formatted_field(file, pkg, "Conflicts");
     pkg_formatted_field(file, pkg, "Status");
+    if (opkg_config->verbose_status_file) {
+        pkg_formatted_field(file, pkg, "Section");
+    }
     pkg_formatted_field(file, pkg, "Essential");
     pkg_formatted_field(file, pkg, "Architecture");
+    if (opkg_config->verbose_status_file) {
+        pkg_formatted_field(file, pkg, "Maintainer");
+        pkg_formatted_field(file, pkg, "MD5sum");
+        pkg_formatted_field(file, pkg, "Size");
+        pkg_formatted_field(file, pkg, "Filename");
+    }
     pkg_formatted_field(file, pkg, "Conffiles");
+    if (opkg_config->verbose_status_file) {
+        pkg_formatted_field(file, pkg, "Source");
+        pkg_formatted_field(file, pkg, "Description");
+    }
     pkg_formatted_field(file, pkg, "Installed-Size");
     pkg_formatted_field(file, pkg, "Installed-Time");
     pkg_formatted_field(file, pkg, "Auto-Installed");
diff --git a/libopkg/pkg_parse.c b/libopkg/pkg_parse.c
index 6dce5cb..302de83 100644
--- a/libopkg/pkg_parse.c
+++ b/libopkg/pkg_parse.c
@@ -118,8 +118,12 @@ int pkg_parse_line(void *ptr, const char *line, uint mask)
     static int reading_conffiles = 0, reading_description = 0;
     int ret = 0;
 
-    /* Exclude globally masked fields. */
-    mask |= opkg_config->pfm;
+    if (opkg_config->verbose_status_file) {
+        mask = 0;
+    } else {
+        /* Exclude globally masked fields. */
+        mask |= opkg_config->pfm;
+    }
 
     /* Flip the semantics of the mask. */
     mask ^= PFM_ALL;
-- 
2.7.4

